### PR TITLE
gnrc_sock_udp: add udp portrange

### PR DESCRIPTION
**Pull request by smlng**
This PR contains parts from #6196 to refine dynamic UDP source port allocation and checks if port already in use, if required.

### CODE REVIEW

File: sys/net/gnrc/sock/udp/gnrc_sock_udp.c [hunk 1]
```diff
@@ -166,23 +212,29 @@ ssize_t sock_udp_send(sock_udp_t *sock, const void *data, size_t len,
...
+            return -EINVAL;
+        }
+        else if (gnrc_af_not_supported(remote->family)) {
+            return -EAFNOSUPPORT;
+        }
+        else if ((sock != NULL) &&
+            (sock->local.netif != SOCK_ADDR_ANY_NETIF) &&
+            (remote->netif != SOCK_ADDR_ANY_NETIF) &&
+            (sock->local.netif != remote->netif)) {
+                return -EINVAL;
```
Comment (miri64) [inline sys/net/gnrc/sock/udp/gnrc_sock_udp.c]: indentation

File: sys/net/gnrc/sock/udp/gnrc_sock_udp.c [hunk 2]
```diff
@@ -166,23 +212,29 @@ ssize_t sock_udp_send(sock_udp_t *sock, const void *data, size_t len,
...
+        }
     }
-    if ((remote == NULL) &&
-        /* sock can't be NULL as per assertion above */
-        (sock->remote.family == AF_UNSPEC)) {
+    else if (sock->remote.family == AF_UNSPEC) {
```
Comment (miri64) [inline sys/net/gnrc/sock/udp/gnrc_sock_udp.c]: The static code analyzer isn't complaining here?

File: sys/net/gnrc/sock/udp/gnrc_sock_udp.c (hunk 1 above)
Comment (smlng) [inline sys/net/gnrc/sock/udp/gnrc_sock_udp.c]: yeah I know, but otherwise its aligns with parentheses above ... will change, but makes it harder to read IMHO

File: sys/net/gnrc/sock/udp/gnrc_sock_udp.c (hunk 2 above)
Comment (smlng) [inline sys/net/gnrc/sock/udp/gnrc_sock_udp.c]: Nope, because of the `assert((sock != NULL) || (remote != NULL))`, i.e., this `else if` only fires if `remote == NULL` and per `assert` then `sock != NULL`.

File: sys/net/gnrc/sock/udp/gnrc_sock_udp.c (hunk 1 above)
Comment (miri64) [inline sys/net/gnrc/sock/udp/gnrc_sock_udp.c]: You could align the parentheses above with the parentheses of the same level in l.230 (that's how I usually do it). This way it's not as hard to read again ;-)

File: sys/include/net/gnrc/udp.h [hunk 3]
```diff
@@ -54,6 +54,13 @@ extern "C" {
 #endif
 
 /**
+ * @brief   Default UDP port range for dynamic source port allocation
+ */
+#define GNRC_UDP_PORTRANGE_MIN (49152U)
```
Comment (OlegHahm) [inline sys/include/net/gnrc/udp.h]: Can we rename "PORTRANGE" to "EPHEMERAL"? Seems less ambiguous to me,

Comment (miri64) [inline sys/include/net/gnrc/udp.h]: Since I always need to google what that term means when you mention it I disagree. The doc sufficiently makes it clear IMHO.

Comment (smlng) [inline sys/include/net/gnrc/udp.h]: mhm, I took the naming from `sysctl -a | grep portrange` on macOS:

```
net.inet.ip.portrange.hilast: 65535
net.inet.ip.portrange.hifirst: 49152
net.inet.ip.portrange.last: 65535
net.inet.ip.portrange.first: 49152
net.inet.ip.portrange.lowlast: 600
net.inet.ip.portrange.lowfirst: 1023
```

IMHO _EPHEMERAL_ isn't that commonly used?

Comment (smlng) [inline sys/include/net/gnrc/udp.h]: you're right, there are lots of port ranges one can possibly define, but at the moment we only need this one :) Because, until now there was not a single portrange for UDP in RIOT ...

That said, I tend to agree that the names could be a little more specific (or speaking), how about `GNRC_UDP_SRC_PORTRANGE_*`? Because `EPHEMERAL` is not very helpful, I agree with @miri64 here.

Comment (OlegHahm) [inline sys/include/net/gnrc/udp.h]: Ephemeral is the term used in the RFC and "Unix Network Programming, Volume 1: Networking APIs: Socket and XTI", Prentice Hall, 1998". Hence, for me it is a well defined term. But "SRC_PORTRANGE" would work for me, if you insist.

**Comment by OlegHahm** [comment]
Btw have you considered https://tools.ietf.org/html/rfc6056 ?

**Comment by smlng** [comment]
@OlegHahm not yet, but thanks for the pointer. 

I consider adding the checks and error mentioned therein. Currently it is possible that `_udp_src_port()` will not return if all ports are used. However, I was kind of _proud_ to get rid of the `random()`, thought it might be unnecessarily expensive on constrained platforms?

**Comment by smlng** [comment]
@OlegHahm and @miri64 adapted parts of RFC 6056, plus naming; please have look.

**Comment by smlng** [comment]
ah the if condition is messed up, wait a moment, will fix

File: sys/net/gnrc/sock/udp/gnrc_sock_udp.c [hunk 4]
```diff
@@ -24,14 +24,66 @@
...
+/**
+ * @brief   returns a UDP port, and checks for reuse if required
+ */
+static uint16_t _udp_dyn_port(sock_udp_t *sock)
+{
+    uint16_t port;
+    uint16_t count = GNRC_UDP_DYN_PORTRANGE_NUM;
+    do {
+        port = GNRC_UDP_DYN_PORTRANGE_MIN +
+               (_udp_dyn_port_next + GNRC_UDP_DYN_PORTRANGE_OFF) % GNRC_UDP_DYN_PORTRANGE_NUM;
```
Comment (miri64) [inline sys/net/gnrc/sock/udp/gnrc_sock_udp.c]: Reading this makes it a little confusing, considering that the variable is named `_udp_dyn_port_next`. Maybe `_udp_dyn_port_cur`.

Comment (miri64) [inline sys/net/gnrc/sock/udp/gnrc_sock_udp.c]: Also, since both variable and these private functions are private and static, consider removing the `_udp` prefix for brevity (but keep the heading `_` to make their private nature clear).

Comment (smlng) [inline sys/net/gnrc/sock/udp/gnrc_sock_udp.c]: naming is taken from rfc 6056, and _next_ fits because it is used for the next port generation.

Comment (miri64) [inline sys/net/gnrc/sock/udp/gnrc_sock_udp.c]: RFC 6056 also uses "ephemeral" ;-P. To me it first looked like you are generating the next port (so the one you take after that)...

Comment (miri64) [inline sys/net/gnrc/sock/udp/gnrc_sock_udp.c]: They are named like this because they are `udp_sock_t`s and not any other `sock` type ;-)

**Comment by smlng** [comment]
@miri64 comments addressed, as I saw fit

File: sys/include/net/gnrc/udp.h [hunk 5]
```diff
@@ -54,6 +54,22 @@ extern "C" {
...
+
+/**
+ * @brief   Offset for next dynamic UDP port
+ *
+ * Currently set to a static (prime) offset, but could be random, too
+ * see https://tools.ietf.org/html/rfc6056#section-3.3.3
+ */
+#define GNRC_UDP_DYN_PORTRANGE_OFF (17U)
+
+/**
```
Comment (miri64) [inline sys/include/net/gnrc/udp.h]: Mhhhh sleeping over it a bit: why are these in `gnrc/udp.h`? Since they are not really used by `gnrc_udp` directly this can be confusing. I would vote putting them either in `gnrc_sock`'s internal header or directly into the c-file.

Comment (smlng) [inline sys/include/net/gnrc/udp.h]: but then, they still have to be moved out of udp.h 😄

Comment (miri64) [inline sys/include/net/gnrc/udp.h]: True. Maybe reflect this in the doc, that this is the *suggested* port range for dynamic allocation in higher layers.

Comment (smlng) [inline sys/include/net/gnrc/udp.h]: mhm, put them into `sys/include/net/sock.h`? Or leave as is, and adapt as soon as TCP is ready? Suggestions?

Comment (miri64) [inline sys/include/net/gnrc/udp.h]: I would NACK that. This is GNRC-specific, so `sys/include/net/sock.h` is definitely the wrong place.

Comment (smlng) [inline sys/include/net/gnrc/udp.h]: OK, but where alternatively, so that tcp can use these defines, too. You don't want to include`udp.h` in sock_tcp 🤔 ?

Comment (miri64) [inline sys/include/net/gnrc/udp.h]: I suggested it before, but how about the internal header `sys/net/gnrc/sock/include/gnrc_sock_internal.h`?

Comment (smlng) [inline sys/include/net/gnrc/udp.h]: yep, sounds good to me (now). Will adapt ...

**Comment by smlng** [comment]
squashed, waiting for murdock and @OlegHahm. 

Thx again for the productive discussion, help and hints!

File: sys/net/gnrc/sock/include/gnrc_sock_internal.h [hunk 6]
```diff
@@ -37,6 +37,22 @@ extern "C" {
 #endif
 
 /**
+ * @brief   Default port range for dynamic source port allocation
+ */
+#define GNRC_SOCK_DYN_PORTRANGE_MIN (49152U)
+#define GNRC_SOCK_DYN_PORTRANGE_MAX (65535U)
```
Comment (OlegHahm) [inline sys/net/gnrc/sock/include/gnrc_sock_internal.h]: But this isn't really GNRC specific, is it?

File: sys/net/gnrc/sock/include/gnrc_sock_internal.h [hunk 7]
```diff
@@ -37,6 +37,22 @@ extern "C" {
...
+#define GNRC_SOCK_DYN_PORTRANGE_NUM (GNRC_SOCK_DYN_PORTRANGE_MAX - GNRC_SOCK_DYN_PORTRANGE_MIN + 1)
+#define GNRC_SOCK_DYN_PORTRANGE_ERR (0)
+
+/**
+ * @brief   Offset for next dynamic port
+ *
+ * Currently set to a static (prime) offset, but could be random, too
+ * see https://tools.ietf.org/html/rfc6056#section-3.3.3
+ */
+#define GNRC_SOCK_DYN_PORTRANGE_OFF (17U)
```
Comment (OlegHahm) [inline sys/net/gnrc/sock/include/gnrc_sock_internal.h]: 23 is a nice prime, too. *j/k*

Does it have to be a prime? Couldn't find that in the RFC. But anyway - if no prime is required, 17 should be as good as any other number (except for 42).

File: sys/net/gnrc/sock/udp/gnrc_sock_udp.c (hunk 2 above)
Comment (OlegHahm) [inline sys/net/gnrc/sock/udp/gnrc_sock_udp.c]: How are this and the changes above related to this PR?

File: sys/net/gnrc/sock/udp/gnrc_sock_udp.c [hunk 8]
```diff
@@ -191,80 +252,50 @@ ssize_t sock_udp_send(sock_udp_t *sock, const void *data, size_t len,
...
-                }
-            }
-#endif
-        }
         memset(&local, 0, sizeof(local));
+        if ((src_port = _get_dyn_port(sock)) == GNRC_SOCK_DYN_PORTRANGE_ERR) {
+            return -EINVAL;
+        }
         if (sock != NULL) {
-            /* bind sock object implicitly */
```
Comment (OlegHahm) [inline sys/net/gnrc/sock/udp/gnrc_sock_udp.c]: Huh?

File: sys/net/gnrc/sock/udp/gnrc_sock_udp.c [hunk 9]
```diff
@@ -191,80 +252,50 @@ ssize_t sock_udp_send(sock_udp_t *sock, const void *data, size_t len,
...
+    /* sock can't be NULL at this point */
     if (remote == NULL) {
-        /* sock can't be NULL at this point */
-        memcpy(&rem, &sock->remote, sizeof(rem));
+        rem = (sock_ip_ep_t *)&sock->remote;
         dst_port = sock->remote.port;
     }
     else {
-        memcpy(&rem, remote, sizeof(rem));
+        rem = (sock_ip_ep_t *)remote;
```
Comment (OlegHahm) [inline sys/net/gnrc/sock/udp/gnrc_sock_udp.c]: Again this seems to be unrelated to me.

File: sys/net/gnrc/sock/include/gnrc_sock_internal.h (hunk 6 above)
Comment (OlegHahm) [inline sys/net/gnrc/sock/include/gnrc_sock_internal.h]: The range of dynamic port numbers are defined by the [IANA](http://www.iana.org/assignments/service-names-port-numbers/service-names-port-numbers.xhtml).

File: sys/net/gnrc/sock/include/gnrc_sock_internal.h (hunk 7 above)
Comment (smlng) [inline sys/net/gnrc/sock/include/gnrc_sock_internal.h]: the RFC actually says it has to be a random component, see code comment. However,  I felt it might not be good to have the overhead of random by default in RIOT

File: sys/net/gnrc/sock/udp/gnrc_sock_udp.c (hunk 8 above)
Comment (smlng) [inline sys/net/gnrc/sock/udp/gnrc_sock_udp.c]: ups, delete by accident

Comment (smlng) [inline sys/net/gnrc/sock/udp/gnrc_sock_udp.c]: on the other hand, I thought the agenda is: don't comment the obvious?

Comment (miri64) [inline sys/net/gnrc/sock/udp/gnrc_sock_udp.c]: Hard to say what's obvious... I wouldn't say it's obvious that the code below is an implicit binding

File: sys/net/gnrc/sock/include/gnrc_sock_internal.h (hunk 6 above)
Comment (smlng) [inline sys/net/gnrc/sock/include/gnrc_sock_internal.h]: the values are not, but @miri64 proposed to put these defines in this header file, hence choosing the prefix `GNRC_SOCK_` seems reasonable to me. However, if we start a naming discussion here, we can unroll every header file in RIOT and check for consistency ...

File: sys/net/gnrc/sock/udp/gnrc_sock_udp.c (hunk 8 above)
Comment (smlng) [inline sys/net/gnrc/sock/udp/gnrc_sock_udp.c]: so I shall re-add?

Comment (OlegHahm) [inline sys/net/gnrc/sock/udp/gnrc_sock_udp.c]: I agree with @miri64. If in doubt: document it! (The change would be unrelated anyway.)

File: sys/net/gnrc/sock/udp/gnrc_sock_udp.c (hunk 9 above)
Comment (OlegHahm) [inline sys/net/gnrc/sock/udp/gnrc_sock_udp.c]: At least it should go into a separate commit. And for my taste: more, but smaller PRs are **less** overhead.

File: sys/net/gnrc/sock/include/gnrc_sock_internal.h (hunk 6 above)
Comment (OlegHahm) [inline sys/net/gnrc/sock/include/gnrc_sock_internal.h]: I don't understand your last comment. My point was: these are well defined numbers and not GNRC specific, so I would define them in a generic (sic!) header file.

File: sys/net/gnrc/sock/include/gnrc_sock_internal.h (hunk 7 above)
Comment (smlng) [inline sys/net/gnrc/sock/include/gnrc_sock_internal.h]: per issue or via mailing list?

File: sys/net/gnrc/sock/udp/gnrc_sock_udp.c (hunk 8 above)
Comment (smlng) [inline sys/net/gnrc/sock/udp/gnrc_sock_udp.c]: ok will re-add

File: sys/net/gnrc/sock/include/gnrc_sock_internal.h (hunk 7 above)
Comment (OlegHahm) [inline sys/net/gnrc/sock/include/gnrc_sock_internal.h]: Maybe the mailing list would be more appropriate indeed.

File: sys/net/gnrc/sock/include/gnrc_sock_internal.h (hunk 6 above)
Comment (smlng) [inline sys/net/gnrc/sock/include/gnrc_sock_internal.h]: I do understand, but again: which header file would be best? Or should we have a iana.h somewhere to organized such official values. In the end we will come to the point where we need IP address ranges or other values assigned by the IANA

Comment (OlegHahm) [inline sys/net/gnrc/sock/include/gnrc_sock_internal.h]: Might be a good idea to have such a header indeed. E.g., protocol implementations could use this header for their default port number.

**Comment by smlng** [comment]
@OlegHahm can you have a final look? Added initial IANA header, as suggested. Any further enhancements in that direction should go in a separate PR, I think?!

**Comment by OlegHahm** [comment]
>  Added initial IANA header, as suggested.

Cool. Can you factorize this in a separate PR?

> Any further enhancements in that direction should go in a separate PR, I think?!

Yupp!

**Comment by smlng** [comment]
>>Added initial IANA header, as suggested.

>Can you factorize this in a separate PR?

@OlegHahm this PR now depends on #6268

**Comment by OlegHahm** [comment]
Needs a rebase now.

**Comment by smlng** [comment]
rebased!

File: sys/net/gnrc/sock/udp/gnrc_sock_udp.c [hunk 10]
```diff
@@ -24,14 +24,70 @@
...
+
+/**
+ * @brief   Checks if a given UDP port is already used by another sock
+ */
+static bool _dyn_port_used(uint16_t port)
+{
+#ifdef MODULE_GNRC_SOCK_CHECK_REUSE
+    for (sock_udp_t *ptr = _udp_socks; ptr != NULL;
+         ptr = (sock_udp_t *)ptr->reg.next) {
+        bool spec_addr = false;
```
Comment (miri64) [inline sys/net/gnrc/sock/udp/gnrc_sock_udp.c]: `spec_addr`? Unspecified/specified might be the wrong terminology here, since it is more like `any` in sockets. As such me and @OlegHahm are a little bit confused about this whole section, since it may well be, that the sock is bound to any address, but the port is set. So it makes no sense to skip entries with all-zero addresses.

Comment (smlng) [inline sys/net/gnrc/sock/udp/gnrc_sock_udp.c]: this wasn't introduced by me, this is _copy-paste_ from the original code. I didn't (want to) tamper with that 😄

File: sys/net/gnrc/sock/udp/gnrc_sock_udp.c [hunk 11]
```diff
@@ -24,14 +24,70 @@
...
+        if ((sock == NULL) || (sock->flags & SOCK_FLAGS_REUSE_EP) ||
+                              !_dyn_port_used(port)) {
+            return port;
+        }
+        --count;
+    } while (count > 0);
+    return GNRC_SOCK_DYN_PORTRANGE_ERR;
+}
+
+/* --------------------------------- public --------------------------------- */
```
Comment (smlng) [inline sys/net/gnrc/sock/udp/gnrc_sock_udp.c]: if that's consesus I will remove such comments, I personally like to use them as little structural helpers.

File: sys/net/gnrc/sock/udp/gnrc_sock_udp.c (hunk 10 above)
Comment (OlegHahm) [inline sys/net/gnrc/sock/udp/gnrc_sock_udp.c]: Okay, I see. In this case I would propose to leave it as is and we open a separate PR to fix this.

File: sys/net/gnrc/sock/udp/gnrc_sock_udp.c [hunk 12]
```diff
@@ -24,14 +24,70 @@
...
+
+/**
+ * @brief   returns a UDP port, and checks for reuse if required
+ *
+ * complies to RFC 6056, see https://tools.ietf.org/html/rfc6056#section-3.3.3
+ */
+static uint16_t _get_dyn_port(sock_udp_t *sock)
+{
+    uint16_t port;
+    uint16_t count = GNRC_SOCK_DYN_PORTRANGE_NUM;
```
Comment (OlegHahm) [inline sys/net/gnrc/sock/udp/gnrc_sock_udp.c]: As this is used as a counter variable, it should be rather `unsigned`.

File: sys/net/gnrc/sock/udp/gnrc_sock_udp.c [hunk 13]
```diff
@@ -24,14 +24,70 @@
...
+    do {
+        port = GNRC_SOCK_DYN_PORTRANGE_MIN +
+               (_dyn_port_next + GNRC_SOCK_DYN_PORTRANGE_OFF) % GNRC_SOCK_DYN_PORTRANGE_NUM;
```
Comment (OlegHahm) [inline sys/net/gnrc/sock/udp/gnrc_sock_udp.c]: Shouldn't it be rather `_dyn_port_next * GNRC_SOCK_DYN_PORTRANGE_OFF` (or `_dyn_port_next++ * GNRC_SOCK_DYN_PORTRANGE_OFF`)?

File: sys/net/gnrc/sock/udp/gnrc_sock_udp.c [hunk 14]
```diff
@@ -24,14 +24,70 @@
...
+        _dyn_port_next++;
+        if ((sock == NULL) || (sock->flags & SOCK_FLAGS_REUSE_EP) ||
+                              !_dyn_port_used(port)) {
```
Comment (OlegHahm) [inline sys/net/gnrc/sock/udp/gnrc_sock_udp.c]: Shouldn't be `_dyn_port_used()` checked in any case, even if `sock == NULL`?

Comment (smlng) [inline sys/net/gnrc/sock/udp/gnrc_sock_udp.c]: @OlegHahm it is, because that's an **or** expression, or what do you suggest/mean?

File: sys/net/gnrc/sock/udp/gnrc_sock_udp.c (hunk 13 above)
Comment (smlng) [inline sys/net/gnrc/sock/udp/gnrc_sock_udp.c]: yeah, you're right!

Comment (smlng) [inline sys/net/gnrc/sock/udp/gnrc_sock_udp.c]: mhm, on the other hand its the algorithm from https://tools.ietf.org/html/rfc6056#section-3.3.3

File: sys/net/gnrc/sock/udp/gnrc_sock_udp.c (hunk 14 above)
Comment (miri64) [inline sys/net/gnrc/sock/udp/gnrc_sock_udp.c]: I guess what @OlegHahm means is, if `(sock == NULL)` lazy-evaluation kicks in and it isn't evaluated ;-).

Comment (smlng) [inline sys/net/gnrc/sock/udp/gnrc_sock_udp.c]: ah, oversaw that one 😞 However, if `sock == NULL` its a onetime (_fire and forget_) usage, i.e. for sending, hence, no check required as no listing will happen. So checking the port is not needed, or do I oversee something (again)?

**Comment by PeterKietzmann** [comment]
@smlng  #6272 was merged

File: sys/net/gnrc/sock/udp/gnrc_sock_udp.c (hunk 12 above)
Comment (smlng) [inline sys/net/gnrc/sock/udp/gnrc_sock_udp.c]: I don't understand, it is unsigned - or do you mean counter should be `int`? However, haven't found anything in the coding conventions to that end ...

File: sys/net/gnrc/sock/udp/gnrc_sock_udp.c (hunk 13 above)
Comment (smlng) [inline sys/net/gnrc/sock/udp/gnrc_sock_udp.c]: I'll go for the multiplication version, though its a bit off the RFC ...

File: sys/net/gnrc/sock/udp/gnrc_sock_udp.c (hunk 9 above)
Comment (smlng) [inline sys/net/gnrc/sock/udp/gnrc_sock_udp.c]: done

**Comment by smlng** [comment]
rebased

File: sys/net/gnrc/sock/udp/gnrc_sock_udp.c (hunk 12 above)
Comment (miri64) [inline sys/net/gnrc/sock/udp/gnrc_sock_udp.c]: The problem isn't one of convention, but of code size: the compiler is better in optimization for generic types (such as `int` or `unsigned`) than with fixed-width types (such as `uint16_t`), since those are tailored to the processors word length. So using `int` or `unsigned` for a counter (unless the width has a purpose, like e.g. a network header field) is always better.

Comment (smlng) [inline sys/net/gnrc/sock/udp/gnrc_sock_udp.c]: ah okay, will adapt

**Comment by PeterKietzmann** [comment]
@OlegHahm do you still request changes?

**Comment by OlegHahm** [comment]
It's still on my review list.

**Comment by smlng** [comment]
rebased

**Comment by PeterKietzmann** [comment]
happy

**Comment by smlng** [comment]
@OlegHahm, @PeterKietzmann: both CI are happy. I think this can go into the release.

**Comment by OlegHahm** [comment]
On it now.
